lua/telescope/previewers/buffer_previewer.lua

```diff
@@ -1,20 +1,94 @@
 local from_entry = require "telescope.from_entry"
 local Path = require "plenary.path"
+local Job = require "plenary.job"
 local utils = require "telescope.utils"
+local strings = require "plenary.strings"
 local putils = require "telescope.previewers.utils"
 local Previewer = require "telescope.previewers.previewer"
+local log = require "telescope.log"
 local conf = require("telescope.config").values
 
 local pfiletype = require "plenary.filetype"
 local pscan = require "plenary.scandir"
 
 local buf_delete = utils.buf_delete
-local defaulter = utils.make_default_callable
 
 local previewers = {}
 
 local ns_previewer = vim.api.nvim_create_namespace "telescope.previewers"
 
+-- any globally configurable opts should be prefixed by `preview_` eg `preview_timeout`
+local function defaulter(f, default_opts)
+  default_opts = default_opts or {}
+  return {
+    new = function(opts)
+      if conf.previewers == false and not opts.show_preview == true then
+        return false
+      end
+      if type(conf.previewers) == "table" then
+        for k, v in pairs(conf.previewers) do
+          local opt = string.format("preview_%s", k)
+          opts[opt] = vim.F.if_nil(opts[opt], v)
+        end
+      end
+      return f(opts)
+    end,
+    __call = function()
+      local ok, err = pcall(f(default_opts))
+      if not ok then
+        error(debug.traceback(err))
+      end
+    end,
+  }
+end
+
+local function set_timeout_message(bufnr, winid, message)
```

Comment:
Ok, it looks like it works sometimes, but seemingly randomly 🤔

Reply:
> You might want to disable scrolling on these buffers.

I don't think that's possible, how would I do that generally? For instance, you could just as well scroll beyond any other previewed buffer right now apart from in the buffer picker. But I think I know how we can limit scrolling.

> Also, for some reason locally the text isn't centered (I love how this display looks btw)

Thanks! Can you reproduce this somehow, I c

Reply:
I looked into scrolling. The way you have to do scrolling so it works as expected always allows you to have the bottom line as the top line. And it seems like we have a bug in there that the last line is always an empty string, that's why can get seemingly outside the buffer. So outside of scope of this PR.

I'll gladly fix centering though if you can give me a reproducible example :)

Reply:
I've managed to find a minimal re-production, but it's kind of flakey as it only works 50% of the time. Also, once it's rendered one way (either right or wrong) it stays that way until you reopen nvim.


1. Create a test file:
```bash
$ yes "This is just a sample line appended  to create a big file." | head -n 1000000 > out.txt
$ ls -lah
$ ls -lah
total 116872
drwxr-xr-x   4 user  wheel   128B 14 Sep 16:31 .
drwxrwxrwt  16 root  wheel   512B 14 Sep 16:24 ..
-rw-r--r--   1 user  wheel   1.4K 14 Sep 16:23 minimal.lua
-rw-r--r--   1 user  wheel    56M 14 Sep 16:18 out.txt
```
2. Open nvim using the minimal config `nvim -nu minimal.lua`
  I used the below minimal config, the main differences are:
    - Using your branch
    - Setting the `filesize_limit`

<details><summary>minimal.lua</summary>

```lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvim/site]]
local package_root = '/tmp/nvim/site/pack'
local install_path = package_root .. '/packer/start/packer.nvim'
local function load_plugins()
  require('packer').startup {
    {
      'wbthomason/packer.nvim',
      {
        'fdschmidt93/telescope.nvim',
        branch = 'feat/preview-timeout',
        requires = {
          'nvim-lua/plenary.nvim',
          { 'nvim-telescope/telescope-fzf-native.nvim', run = 'make' },
        },
      },
      -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. '/plugin/packer_compiled.lua',
      display = { non_interactive = true },
    },
  }
end
_G.load_config = function()
  require('telescope').setup({
      defaults = {
          previewers = {
              filesize_limit = 150
          }
      }
  })
  require('telescope').load_extension('fzf')
  -- ADD INIT.LUA SETTINGS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
end
if vim.fn.isdirectory(install_path) == 0 then
  print("Installing Telescope and dependencies.")
  vim.fn.system { 'git', 'clone', '--depth=1', 'https://github.com/wbthomason/packer.nvim', install_path }
end
load_plugins()
require('packer').sync()
vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]]
```
</details>

3. Run `:Telescope find_files` and navigate to the file
4. Try it a couple times until it works 🤷

Reply:
Cannot reproduce :(

https://user-images.githubusercontent.com/39233597/133324669-c9c91bf3-d3c2-4a97-881d-07da3f7e05d8.mp4

Reply:
I have closed and re-opened nvim with minimal init opening `out.txt` and in several other folders with large files (txt files, binaries, etc). Still doesn't happen for me. :/

Reply:
Ah, I've also realized `filesize_limit` was set wrongly for me with your minimal init lua because we've changed to preview (for the time being). Now it just times out (also as intended though).

Race condition sounds very unlikely, the functions as a whole are adequately scheduled and if we set the message, nothing else should affect the buffer. We need `vim.api.nvim_win_get_width` to adequately fill background of preview and then just lay over the message with an extmark on top of that.

Sorry, out of my depth here as well. Hopefully someone else can also repro to help pin this down?